File: exercises/python/broken.py
"""
Intentionally broken Python code for practicing LSP diagnostics.

Keybindings to practice:
  [d / ]d    - navigate between diagnostics
  <leader>ld - list all diagnostics (Trouble)
  <leader>ca - code actions (quick fix)
  K          - hover documentation
"""

import os
import json  # noqa: F401 - unused import (intentional)
from typing import Optional
from collections import OrderedDict  # noqa: F401 - unused import


# ERROR: missing import for 'math' module
def calculate_hypotenuse(a: float, b: float) -> float:
    """Calculate the hypotenuse of a right triangle."""
    return math.sqrt(a ** 2 + b ** 2)


# ERROR: undefined variable 'config'
def get_database_url() -> str:
    """Get database URL from config."""
    host = config["database"]["host"]
    port = config["database"]["port"]
    return f"postgresql://{host}:{port}/mydb"


# ERROR: wrong argument count
def greet(name: str, greeting: str, punctuation: str) -> str:
    """Generate a greeting message."""
    return f"{greeting}, {name}{punctuation}"

message = greet("Alice")  # missing 2 required arguments


# ERROR: type mismatch
def add_numbers(a: int, b: int) -> int:
    """Add two integers."""
    return a + b

result: int = add_numbers("hello", "world")  # wrong types


# ERROR: attribute does not exist
class User:
    def __init__(self, name: str, email: str) -> None:
        self.name = name
        self.email = email

user = User("Jan", "jan@example.com")
print(user.phone)  # 'User' has no attribute 'phone'


# ERROR: incompatible return type
def get_active_users() -> list[str]:
    """Should return a list of strings but returns None."""
    users = ["Alice", "Bob", "Charlie"]
    if len(users) > 0:
        return None  # should return users


# ERROR: unreachable code
def process_data(data: Optional[str]) -> str:
    """Process data string."""
    if data is None:
        raise ValueError("Data cannot be None")
        print("This line is unreachable")  # unreachable
    return data.strip()


# ERROR: variable used before assignment
def compute_average(numbers: list[float]) -> float:
    """Compute average of a list."""
    if len(numbers) > 0:
        total = sum(numbers)
    return total / len(numbers)  # 'total' possibly unbound


# ERROR: wrong dict key type
scores: dict[str, int] = {"alice": 95, "bob": 87}
value = scores[42]  # key should be str, not int

File: exercises/typescript/broken.ts
/**
 * Intentionally broken TypeScript code for LSP diagnostics practice.
 *
 * Keybindings to practice:
 *   [d / ]d    - navigate between diagnostics
 *   <leader>ld - list diagnostics (Trouble)
 *   <leader>ca - code actions
 *   K          - hover for type info
 */

// ERROR: unused imports
import { readFileSync, writeFileSync, existsSync } from "fs";
import { join, resolve, basename } from "path";

// ERROR: duplicate identifier
interface User {
  id: number;
  name: string;
  email: string;
}

// ERROR: missing required properties
const admin: User = {
  id: 1,
  name: "Admin",
  // missing 'email' property
};

// ERROR: wrong generic type argument
function getFirst<T>(items: T[]): T {
  return items[0];
}

const numbers: string = getFirst<number>([1, 2, 3]); // Type 'number' is not assignable to 'string'

// ERROR: property does not exist on type
interface Product {
  id: number;
  name: string;
  price: number;
}

function formatProduct(product: Product): string {
  return `${product.name} - ${product.description}`; // 'description' does not exist
}

// ERROR: argument type mismatch
function calculateTotal(prices: number[]): number {
  return prices.reduce((sum, price) => sum + price, 0);
}

const total = calculateTotal("not an array"); // string is not number[]

// ERROR: unreachable code
function processStatus(status: "active" | "inactive"): string {
  switch (status) {
    case "active":
      return "User is active";
    case "inactive":
      return "User is inactive";
  }
  console.log("This is unreachable"); // unreachable code
  return "unknown";
}

// ERROR: async/await mismatch
function fetchData(): string {
  const response = await fetch("https://api.example.com/data"); // await in non-async
  return response.json();
}

// ERROR: wrong return type
function divide(a: number, b: number): number {
  if (b === 0) {
    return "Cannot divide by zero"; // string is not number
  }
  return a / b;
}

// ERROR: using private member
class BankAccount {
  private balance: number;

  constructor(initial: number) {
    this.balance = initial;
  }

  getBalance(): number {
    return this.balance;
  }
}

const account = new BankAccount(1000);
console.log(account.balance); // 'balance' is private

// ERROR: type narrowing issue
function processValue(value: string | number | null): string {
  // forgot to handle null case
  return value.toString(); // Object is possibly 'null'
}

// ERROR: readonly assignment
interface Config {
  readonly apiUrl: string;
  readonly timeout: number;
}

const config: Config = { apiUrl: "https://api.example.com", timeout: 5000 };
config.apiUrl = "https://other.api.com"; // Cannot assign to readonly property
